qkernel/src/interrupt/aarch64/mod.rs

```diff
@@ -1,6 +1,309 @@
+
+    match ec {
+        EsrDefs::EC_DATA_ABORT => {
+            let far = GetFarEL1();
+            MemAbortKernel(ptregs_addr, esr, far, false);
+        },
+        EsrDefs::EC_INSN_ABORT => {
+            let far = GetFarEL1();
+            MemAbortKernel(ptregs_addr, esr, far, true);
+        },
+        _ => {
+            panic!("unhandled sync exception from el1: {}\n", ec);
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn exception_handler_el1h_irq(ptregs_addr:usize){
+    return exception_handler_unhandled(ptregs_addr, 5);
+}
+#[no_mangle]
+pub extern "C" fn exception_handler_el1h_fiq(ptregs_addr:usize){
+    return exception_handler_unhandled(ptregs_addr, 6);
+}
+#[no_mangle]
+pub extern "C" fn exception_handler_el1h_serror(ptregs_addr:usize){
+    return exception_handler_unhandled(ptregs_addr, 7);
+}
+
+// TODO add parameter to the handler
+// TODO implement el0_64_sync handler for syscalls
+#[no_mangle]
+pub extern "C" fn exception_handler_el0_sync(ptregs_addr:usize){
+    let esr = GetEsrEL1();
+    let ec = EsrDefs::GetExceptionFromESR(esr);
+    if ptregs_addr == 0 {
+        panic!("exception frame is null pointer\n")
+    }
+    match ec {
+        EsrDefs::EC_SVC => {
+            // arm64 linux syscall calling convention
+            // TODO maybe there is a better/safer way of this pointer cast
+            let ctx_p = ptregs_addr as *mut PtRegs;
+            let ctx_p = ctx_p.cast::<PtRegs>();
+            let ctx = unsafe { &mut *ctx_p };
+            // syscall number from w8
+            let call_no = ctx.regs[8] as u32;
+            let arg0 = ctx.regs[0];
+            let arg1 = ctx.regs[1];
+            let arg2 = ctx.regs[2];
+            let arg3 = ctx.regs[3];
+            let arg4 = ctx.regs[4];
+            let arg5 = ctx.regs[5];
+            // write syscall ret to x0
+            ctx.regs[0] = syscall_dispatch_aarch64(call_no,arg0,arg1,arg2,arg3,arg4,arg5);
+            // TODO do we need to write the "second ret val" back to x1?
+        },
+        EsrDefs::EC_DATA_ABORT_L => {
+            let far = GetFarEL1();
+            MemAbortUser(ptregs_addr, esr, far, false);
+        },
+        EsrDefs::EC_INSN_ABORT_L => {
+            let far = GetFarEL1();
+            MemAbortUser(ptregs_addr, esr, far, true);
+        },
+        _ => {
+            panic!("unhandled sync exception from el0: {}\n", ec);
+        }
+        // TODO (default case) for a unhandled exception from user,
+        // the kill the user process instead of panicing
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn exception_handler_el0_irq(){return;}
+#[no_mangle]
+pub extern "C" fn exception_handler_el0_fiq(){return;}
+#[no_mangle]
+pub extern "C" fn exception_handler_el0_serror(){return;}
+
+#[inline]
+pub fn GetFaultAccessType(esr:u64, is_exe:bool) -> AccessType{
+    if is_exe {
+        return AccessType(MmapProt::PROT_EXEC);
+    }
+    if !EsrDefs::IsCM(esr) && EsrDefs::IsWnR(esr){
```

Comment:
@shrik3 I am not sure about the logic here. If `CM==1 -> IsCM==T -> !IsCM==F -> expression equals F.`
I would suggest `IsCM() || IsWnR()`.